### PR TITLE
composite-checkout: Update domain contact validation

### PR DESCRIPTION
**Pull request by nbloomf**
#### Changes proposed in this Pull Request

* Update domain contact validation to use the new async strategy introduced in #39636. Also depends on #39837.

#### Testing instructions

Enter checkout with a domain and proceed to the contact details step. Try to get past it without filling out the form properly. For example:

- Click 'continue' without filling in any data
- Only fill in the name
- Put nonnumeric characters in the phone number field
- Put a phone number with no area code in the phone number field
- US, UK, and EU countries have specialized address fields - try to break these too.

### CODE REVIEW

File: client/my-sites/checkout/checkout/composite-checkout.js
Comment (sirbrillig) [inline client/my-sites/checkout/checkout/composite-checkout.js]: Below, when `domainContactValidationCallback` is being called, it has an `await` before it. However, I'm pretty sure that this function does not return a Promise. The default, `Undocumented.prototype.validateDomainContactInformation`, passes a callback argument to `wpcom.req.post` which I believe makes it use the callback instead of a Promise.

Since we probably need to wait for this to complete before continuing, you could create a Promise here instead by changing the return value of this function to `return new Promise(resolve => { ... } )` and then in the callback below, after `applyDomainContactValidationResults`, call `resolve()`.

**Comment by fditrapani** [comment]
Thanks for getting this going. I took this for a test and here's what I found:

- When I press continue, and have nothing filled in, I see a message show up like this:
 image

It would be better if we showed the errors on each field rather than showing a global notice because I don't know which fields to fix when they aren't highlighted. 

- After an initial submission, all the errors show up when I type in one field: 
image

- As I type in a field, the error message remains. It would be nice to clear the error as soon as it validates. In the case of a blank field, the message should go away right after you type something (if it's valid :D). 

- When I press continue, it clears the global notice and leaves the field errors, but doesn't continue. I have to press continue again. Can we clear all the errors and continue at the same time?
- When I go back to edit the step, the field errors are still showing:
image
- Auto complete seems to be working now :)
- I still seem to be able to submit invalid phone numbers and postal codes and sometimes it validates. It's not applying the validation consistently.
- Rather than showing a global notice inline as you have, can we show it as one of the system notices we use for the coupons? this should help folks on mobile. The message can read "We could not validate your contact information. Please review and update all the highlighted fields."

**Review by sirbrillig — COMMENTED** [review]
I approve all this work; it works for me! Nice job. Just not actually marking it approved until #39837 is merged and we rebase this on top of master.

**Review by fditrapani — APPROVED** [review]
This is looking great @nbloomf! It works much better now: 
- The browser's auto-complete works flawlessly.
- The errors show and hide as expected.
- The notice works well to communicate there's an error.

Here are a couple comments that aren't blockers and we can deal with in separate PRs:
- There doesn't seem to be client side validation on the fields. If I enter invalid info, or start typing and then empty a field, we don't throw an error until you submit the form. It's a micro optimization but can help people complete the form quicker. 
- When I submit the form on a mobile screen, it's possible that the error could appear at the top of the form and out of view. I'm wondering whether we should take people to the top of the form when they submit and there's an error. Similarly, they would have to scroll down to see the error if the error was on one of the fields closer to the button. What do you think? 
- When I submit an updated form, the global error notification doesn't go away.
- On desktop, the errors have very little space and break out into multiple short lines. We can reduce this impact by removing the icon and padding, and adding a red border around the field for more emphasis. Is it possible to isolate this change to the composite checkout?

**Before**
image
 
**After**
image

**Comment by nbloomf** [comment]
> Caution: This PR has changes that must be merged to WordPress.com

I am 98% sure this happened because I rebased and force-pushed this before updating the PR to merge into master (instead of #39837).

**Comment by sirreal** [comment]
> > Caution: This PR has changes that must be merged to WordPress.com
> 
> I am 98% sure this happened because I rebased and force-pushed this before updating the PR to merge into master (instead of #39837).

Seems likely. I've removed a slew of code owners reviewer requests as well 🤷‍♂ Feel free to re-add if they're desired.

**Review by sirbrillig — APPROVED** [review]
Still works after the rebase! Nice work.

**Comment by noahtallen** [comment]
> I am 98% sure this happened because I rebased and force-pushed this before updating the PR to merge into master

Makes sense! That build triggers if it detects changes to `apps/full-site-editing`

**Pull request merged**
